compiler/pipes/final-check.cpp

```diff
@@ -408,10 +411,50 @@ VertexPtr FinalCheckPass::on_enter_vertex(VertexPtr vertex) {
     check_null_usage_in_binary_operations(binary_vertex);
   }
 
+  if (vertex->type() == op_index) {
+    check_indexing(vertex);
+  }
+
   //TODO: may be this should be moved to tinf_check
   return vertex;
 }
 
+void FinalCheckPass::check_indexing(VertexPtr vertex) {
```

Comment:
It wold be better to accept `VertexAdaptor<op_index>` here

Reply:
Fixed.